artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/RefCountMessage.java

```diff
@@ -49,6 +93,28 @@ public int getDurableCount() {
       return DURABLE_REF_COUNT_UPDATER.get(this);
    }
 
+   /** Sub classes may mark fired=true when they were explicitly called.
+    * E.g large message removed the file upon cancellation */
+   protected void markFired() {
+      fired = true;
```

Comment:
if the system called the fire method directly, say, a cancellation from a bad producer, there's no need to call the fire event again on countDown... (which might not actually happen since the message is going to be GCed).


I need the variable in the status as well as I can't have access to the message from the status.

Reply:
So is the idea that during the 'normal' countDown it marks the 'outer fired' variable, reaches out somewhere else in firing releaseComplete(), then that other places [maybe] reaches back in here eventually and calls 'markFired()' and doubly-sets the 'outer fired' variable and ALSO sets the separate 'debugStatus fired' variable to _really_ mark the message as cleaned up? The cleaner only seems to look at the latter variable when deciding whether the message was cleaned up or not, but only the markFired() 'direct call' is setting that. I looked at/thought about this code for 30mins earlier and am asking questions as I still dont understand what its expected to do normally. Its pretty impenetrable as-is. If I wasnt seeing the diff I probably wouldnt even have noticed the 2 different 'fired' variables, with one being used in 2 different ways by the two routes through the code.

Reply:
The first variable is used to control the releaseResources...

The second variable is used to control if the resource leaked or not.


I'm renaming it to leaked.




and I can't read anything from RefCountMessage coming from RefCountDebugState (I just renamed it BTW).

Having a reference back to RefCountMessage would make it impossible to release the reference. and I don't want to complicate things with a WeakReference... I would rather duplicate the variable... in this case I renamed it to leaked to make a distinction.